src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs

```diff
@@ -362,107 +407,78 @@ private static Version GetMSBuildVersionInPath()
-            }
-            else
-            {
-                // append ".0" if the userVersion is a number
-                string userVersionString = userVersion;
-                int unused;
-
-                if (int.TryParse(userVersion, out unused))
+                // then match by Major & Minor version numbers. And we want an actual parsing of t.ToolsVersion,
+                // without the safe fallback to 0.0 built into t.ParsedToolsVersion.
+                Version parsedUserVersion;
+                Version parsedToolsVersion;
+                if (Version.TryParse(userVersionString, out parsedUserVersion) &&
+                    Version.TryParse(t.ToolsVersion, out parsedToolsVersion))
                 {
-                    userVersionString = userVersion + ".0";
+                    return parsedToolsVersion.Major == parsedUserVersion.Major &&
+                        parsedToolsVersion.Minor == parsedUserVersion.Minor;
                 }
 
-                Version ver;
-                bool hasNumericVersion = Version.TryParse(userVersionString, out ver);
-
-                var selectedToolset = installedToolsets.FirstOrDefault(
-                toolset =>
-                {
-                    // first match by string comparison
-                    if (string.Equals(userVersionString, toolset.ToolsVersion, StringComparison.OrdinalIgnoreCase))
-                    {
-                        return true;
-                    }
+                return false;
+            });
 
-                    // then match by Major & Minor version numbers.
-                    Version toolsVersion;
-                    if (hasNumericVersion && Version.TryParse(toolset.ToolsVersion, out toolsVersion))
-                    {
-                        return (toolsVersion.Major == ver.Major &&
-                            toolsVersion.Minor == ver.Minor);
-                    }
+            if (selectedToolset == null)
+            {
+                var message = string.Format(
+                    CultureInfo.CurrentCulture,
+                    LocalizedResourceManager.GetString(
+                        nameof(NuGetResources.Error_CannotFindMsbuild)),
+                    userVersion);
 
-                    return false;
-                });
+                throw new CommandLineException(message);
+            }
 
-                if (selectedToolset == null)
-                {
-                    var message = string.Format(
-                        CultureInfo.CurrentCulture,
-                        LocalizedResourceManager.GetString(
-                            nameof(NuGetResources.Error_CannotFindMsbuild)),
-                        userVersion);
+            return selectedToolset;
+        }
 
-                    throw new CommandLineException(message);
-                }
+        private static void LogToolsetToConsole(IConsole console, MsBuildToolsetEx toolset)
+        {
+            if (console == null)
```

Comment:
Actually, that will break us if console is null, which is a valid scenario per previous logic.